### ISSUE TITLE
Usage with TypeScript

### ISSUE THREAD

**Issue by lazarljubenovic** [csstree/csstree#88]
Are there plans to include .d.ts files so this library can be used with TypeScript in a type-safe manner?

**Comment by lahmatiy** [comment]
Yep, I planned to add `.d.ts` to the project. @smelukov [worked on TypeScript definition](https://github.com/smelukov/csstree/tree/dts-file/types), but didn't finish. It will be great if someone finishes work.

**Comment by lazarljubenovic** [comment]
Have you thought about migrating to TS in source code? Writing declaration files is boring, error-prone, and needs manual updates every time source code changes. Rewriting the code into TypeScript would make it much more maintainable and easier for refactoring.

Would you accept such a PR?

**Comment by GrandSchtroumpf** [comment]
Working with Typescript is a huge plus, and it's much easier to share API with others. I would recommend that to if possible.

**Comment by DeDuckProject** [comment]
I actually started writing one and then notices that there's a [@types/css-tree](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/014c05022b2f2044f92d21a6a0c9b035e67289c1/types/css-tree/index.d.ts) so there you go 💃🏼

**Comment by kkirby** [comment]
@DeDuckProject Those types aren't exactly the best. For example:

```
export interface SelectorList extends CssNodeCommon {
    type: 'SelectorList';
    children: List<CssNode>;
}
```

should really be:

```
export interface SelectorList extends CssNodeCommon {
    type: 'SelectorList';
    children: List<Selector>;
}
```

**Comment by wartab** [comment]
Yeah, those typings are not working right. It's sadly a huge amount of work, for example when it comes to having typings correct when using the visit option in walk().

**Comment by lazarljubenovic** [comment]
A year later I still believe this module would greatly benefit from being completely rewritten in TypeScript. It would mostly be adding types to existing code, occasionally refactoring tiny bits for more complex scenarios. It wouldn't introduce any breaking changes.

**Comment by 43081j** [comment]
I started converting it to typescript out of interest.

The current source is very... non-typescript friendly.

You have so many occurrences of applying arguments via `Function.prototype.apply` to bind a custom `this` that nobody has any clue what anything is at any given time.

For example, a parser would be an instance of a `Parser` class, maybe? But no, it is an instance of `TokenStream` _with a load of methods dynamically merged in_.

Even the `readSequence` function, it takes a `recognizer`. What is a "recognizer"? Oh and this function, `readSequence`, again somehow calls `this` (`this.tokenType` for example). 

This pattern of binding everything at run-time to some custom context is why it'll be an insane pain to convert to typescript. All those functions need changing to accept a parser as a parameter rather than doing some dynamic-this magic.

I might keep messing around with my branch but its a deep rabbit hole to go down.

**Comment by lazarljubenovic** [comment]
I've been fiddling around with a TypeScript on and off for some time now. The main issue for me seems to be abusing the CommonJS exports structure.  I have troubles figuring out if an object is actually needed, or is the object created so that several symbols can be exported. There's quite a few properties which have been added on functions as well, which doesn't seem to have any use other than "too lazy to import that as well so let's just use it from the function".

**Comment by 43081j** [comment]
I think to convert this project to typescript it would make a lot of sense to refactor a little. 

I would suggest any visitor-like function is passed a "context" which holds a reference to the parser. Instead of what it does now: bind to the parser when it is called. 

This way you can test all the individual functions standalone too. 

I spent an hour playing around with converting it and I think the most sensible place to start would be to define the fundamental types (e.g. `Node`). 

From there you can start strongly typing most of the functions. 

There's a few places where functions have properties like you say. They should be split up into separate exports. 

And most of the large objects should become classes, e.g. `Parser`, `Lexer`, etc. 

It's a fair amount of refactoring though so it'd be a big job. None of the types are difficult really, theres just a lot of them and not all are obvious from the source.

**Comment by maltenuhn** [comment]
Would this project be open to us offering a bounty? We're making pretty heavy use of this amazing project, and having better support for (and ideally a version written in) Typescript would help us a lot - but we're not experts in this codebase.

**Comment by kkirby** [comment]
@maltenuhn I certainly wouldn't argue. :)

**Comment by maltenuhn** [comment]
@kkirby fantastic! Do you know how to best go about that - specifically, Is it something the csstree owners would need to approve / post / set up?

**Comment by kkirby** [comment]
@maltenuhn https://www.bountysource.com/issues/60509916-usage-with-typescript

**Comment by peterbe** [comment]
👋 There are bugs in https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/css-tree/index.d.ts
For example, `isEmpty` on the `List` class is not a method. It's a getter. 
So instead of:
```ts
export class List<TData> {
    ...
    isEmpty(): boolean;
```
I think it needs to be:
```ts
export class List<TData> {
    ...
    readonly isEmpty: boolean;
```

right?
I've never contributed to `DefinitelyTyped` but would it be an option to at least copy what they started over there and put it into *this* repo so we have some hope of keeping it up-to-date @lahmatiy ?

**Comment by peterbe** [comment]
For what it's worth, my `eslint` forced me to jump through hoops. So I ended up with this:
```ts
// `node.prelude.children.isEmpty` is a getter, not a method.
// Pretty sure it's a bug in
// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/css-tree/index.d.ts
// @ts-expect-error But in DefinitelyTyped for css-tree
if (node.prelude.children.isEmpty as boolean) {
  list.remove(item);
}
```

Now eslint, `tsc` and Node are finally all happy.

**Comment by scripthunter7** [comment]
Reference: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/62536
